firebase-sessions/src/main/kotlin/com/google/firebase/sessions/api/FirebaseSessionsDependencies.kt

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.firebase.sessions.api
+
+import androidx.annotation.VisibleForTesting
+import java.util.Collections.synchronizedMap
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+/**
+ * [FirebaseSessionsDependencies] determines when a dependent SDK is installed in the app. The
+ * Sessions SDK uses this to figure out which dependencies to wait for to getting the data
+ * collection state. This is thread safe.
+ *
+ * This is important because the Sessions SDK starts up before dependent SDKs.
+ */
+object FirebaseSessionsDependencies {
+  private val dependencies = synchronizedMap(mutableMapOf<SessionSubscriber.Name, Dependency>())
+
+  /**
+   * Add a subscriber as a dependency to the Sessions SDK. Every dependency must register itself, or
+   * the Sessions SDK will never generate a session.
+   */
+  fun addDependency(subscriberName: SessionSubscriber.Name) {
+    if (dependencies.containsKey(subscriberName)) {
+      throw IllegalArgumentException("Dependency $subscriberName already added.")
+    }
+
+    // The dependency is locked until the subscriber registers itself.
+    dependencies[subscriberName] = Dependency(Mutex(locked = true))
+  }
+
+  /** Register and unlock the subscriber. This must be called before [getSubscribers] can return. */
+  internal fun register(subscriber: SessionSubscriber) {
+    val subscriberName = subscriber.sessionSubscriberName
+    val dependency = getDependency(subscriberName)
+
+    dependency.subscriber?.run {
+      throw IllegalArgumentException("Subscriber $subscriberName already registered.")
```

Comment:
I'm thinking that re-register does not have to throw an exception either. Again a similar rationale - It is not a breaking situation for the app/developers.

Reply:
I made both `addDependency` and `register` not throw, they will log a debug message and return.